sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala

```diff
@@ -346,36 +319,50 @@ trait ProverInterpreter extends Interpreter with AttributionCore {
     t.withPolynomial(q).copy(children=newChildren)
     */
 
-
-
     case su: UnprovenSchnorr =>
-      if (su.simulated) {
-        // Step 5 (simulated leaf -- complete the simulation)
-        assert(su.challengeOpt.isDefined)
-        val (fm, sm) = DLogInteractiveProver.simulate(su.proposition, su.challengeOpt.get)
-        UncheckedSchnorr(su.proposition, Some(fm), su.challengeOpt.get, sm)
-      } else {
-        // Step 6 (real leaf -- compute the commitment a)
-        val (r, commitment) = DLogInteractiveProver.firstMessage(su.proposition)
-        su.copy(commitmentOpt = Some(commitment), randomnessOpt = Some(r))
+      //Steps 5 & 6: pull out commitment from the hints bag, otherwise, compute the commitment(if the node is real),
+      // or simulate it (if the node is simulated)
+
+      // Step 6 (real leaf -- compute the commitment a or take it from the hints bag)
+      hintsBag.commitments.find(_.image == su.proposition).map { cmtHint =>
+        su.copy(commitmentOpt = Some(cmtHint.commitment.asInstanceOf[FirstDLogProverMessage]))
+      }.getOrElse {
+        if (su.simulated) {
+          // Step 5 (simulated leaf -- complete the simulation)
+          assert(su.challengeOpt.isDefined)
+          val (fm, sm) = DLogInteractiveProver.simulate(su.proposition, su.challengeOpt.get)
+          UncheckedSchnorr(su.proposition, Some(fm), su.challengeOpt.get, sm)
+        } else {
+          // Step 6 -- compute the commitment
+          val (r, commitment) = DLogInteractiveProver.firstMessage()
+          su.copy(commitmentOpt = Some(commitment), randomnessOpt = Some(r))
+        }
       }
 
     case dhu: UnprovenDiffieHellmanTuple =>
-      if (dhu.simulated) {
-        // Step 5 (simulated leaf -- complete the simulation)
-        assert(dhu.challengeOpt.isDefined)
-        val (fm, sm) = DiffieHellmanTupleInteractiveProver.simulate(dhu.proposition, dhu.challengeOpt.get)
-        UncheckedDiffieHellmanTuple(dhu.proposition, Some(fm), dhu.challengeOpt.get, sm)
-      } else {
-        // Step 6 (real leaf -- compute the commitment a)
-        val (r, fm) = DiffieHellmanTupleInteractiveProver.firstMessage(dhu.proposition)
-        dhu.copy(commitmentOpt = Some(fm), randomnessOpt = Some(r))
-      }
+       //Steps 5 & 6: pull out commitment from the hints bag, otherwise, compute the commitment(if the node is real),
+       // or simulate it (if the node is simulated)
+
+        // Step 6 (real leaf -- compute the commitment a or take it from the hints bag)
+        hintsBag.commitments.find(_.image == dhu.proposition).map { proof =>
```

Comment:
Any reason its called `proof` here and `cmtHint` in `UnprovenSchnorr` case?

Reply:
sloppiness ) renamed